src/dynamics/ergodic/add_circle.lean

```diff
@@ -0,0 +1,102 @@
+/-
+Copyright (c) 2022 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import measure_theory.group.add_circle
+import dynamics.ergodic.ergodic
+import measure_theory.covering.density_theorem
+import data.set.pointwise.iterate
+
+/-!
+# Ergodic maps of the additive circle
+
+This file contains proofs of ergodicity for maps of the additive circle.
+
+# Main definitions:
+
+ * `add_circle.ergodic_zsmul`: given `n : ℤ` such that `1 < |n|`, the self map `y ↦ n • y` on
+   the additive circle is ergodic (wrt the Haar measure).
+
+# TODO
+
+ * The map `y ↦ n • y + x` is ergodic for any `x` (and `1 < |n|`).
+
+-/
+
+open set function measure_theory measure_theory.measure filter metric
+open_locale measure_theory nnreal ennreal topological_space pointwise
+
+namespace add_circle
+
+variables {T : ℝ} [hT : fact (0 < T)]
+include hT
+
+/-- If a null-measurable subset `s` of the circle is invariant under rotation by an arbitrarily
+small rational angle, then it must be almost empty or almost full. -/
```

Comment:
How does `hu₂` fit into this docstring? It's possible this is very clear and I'm just not familiar with the terminology...

Reply:
An element `g` has finite iff `0 < add_order_of g`.

Assume `T = 1` for simplicity. Then, the points of finite order on the circle are the points with a "rational angle", i.e., they have a representative of the form `m / n` for `m : ℤ` and `n : ℕ`. If this is in lowest terms then `add_order_of ↑(m/n) = n`.

So the hypothesis `hu₂` says that we have a sequence of rational points which when we add (aka "rotate" if we allow this terminology) then the set is preserved but `n → ∞`.

(I've actually just reworded the comment after writing this: what is proved is actually a bit stronger than the original wording.)